File: app/src/main/java/com/example/recipeclipper/data/model/RecipeDraft.kt
package com.example.recipeclipper.data.model

/**
 * What the Edit screen saves (#29): a recipe's content as the user typed it, before it is
 * stored. Pure, like the parsers: text in, data out. Nothing here is guessed or converted;
 * the user's words are kept as written, only trimmed.
 */
data class RecipeDraft(
    val name: String = "",
    val yield: String = "",
    val prepTime: String = "",
    val cookTime: String = "",
    val totalTime: String = "",
    /** One ingredient per line, as typed in the ingredients box. */
    val ingredientsText: String = "",
    /** One step per line, as typed in the steps box. */
    val instructionsText: String = "",
    val image: String = ""
) {
    val ingredients: List<String> get() = lines(ingredientsText)
    val instructions: List<String> get() = lines(instructionsText)

    /** The parsers' rule for a recipe: a name, plus ingredients or steps. */
    val isValid: Boolean
        get() = name.isNotBlank() && (ingredients.isNotEmpty() || instructions.isNotEmpty())

    /**
     * [base]'s identity and user state with this draft's content. A blank optional field is
     * absent (null), never an empty string.
     */
    fun applyTo(base: Recipe): Recipe = base.copy(
        name = name.trim(),
        image = image.trim().ifEmpty { null },
        ingredients = ingredients,
        instructions = instructions,
        prepTime = prepTime.trim().ifEmpty { null },
        cookTime = cookTime.trim().ifEmpty { null },
        totalTime = totalTime.trim().ifEmpty { null },
        yield = yield.trim().ifEmpty { null }
    )

    companion object {
        /** A recipe's content as the Edit screen first shows it. */
        fun of(recipe: Recipe) = RecipeDraft(
            name = recipe.name,
            yield = recipe.yield.orEmpty(),
            prepTime = recipe.prepTime.orEmpty(),
            cookTime = recipe.cookTime.orEmpty(),
            totalTime = recipe.totalTime.orEmpty(),
            ingredientsText = recipe.ingredients.joinToString("\n"),
            instructionsText = recipe.instructions.joinToString("\n"),
            image = recipe.image.orEmpty()
        )

        /** One entry per non-blank line, trimmed. */
        fun lines(text: String): List<String> =
            text.split('\n').map { it.trim() }.filter { it.isNotEmpty() }
    }
}

File: ios/RecipeClipper/Data/Model/RecipeDraft.swift
import Foundation

/// What the Edit screen saves (#29): a recipe's content as the user typed it, before it is
/// stored. Pure, like the parsers (Android's RecipeDraft.kt). Nothing here is guessed or
/// converted; the user's words are kept as written, only trimmed.
struct RecipeDraft: Equatable {
    var name = ""
    var yield = ""
    var prepTime = ""
    var cookTime = ""
    var totalTime = ""
    /// One ingredient per line, as typed in the ingredients box.
    var ingredientsText = ""
    /// One step per line, as typed in the steps box.
    var instructionsText = ""
    var image = ""

    var ingredients: [String] { Self.lines(ingredientsText) }
    var instructions: [String] { Self.lines(instructionsText) }

    /// The parsers' rule for a recipe: a name, plus ingredients or steps.
    var isValid: Bool {
        !Self.trimmed(name).isEmpty && (!ingredients.isEmpty || !instructions.isEmpty)
    }

    /// `base`'s identity and user state with this draft's content. A blank optional field is
    /// absent (nil), never an empty string.
    func apply(to base: Recipe) -> Recipe {
        var recipe = base
        recipe.name = Self.trimmed(name)
        recipe.image = Self.optional(image)
        recipe.ingredients = ingredients
        recipe.instructions = instructions
        recipe.prepTime = Self.optional(prepTime)
        recipe.cookTime = Self.optional(cookTime)
        recipe.totalTime = Self.optional(totalTime)
        recipe.yield = Self.optional(yield)
        return recipe
    }

    /// A recipe's content as the Edit screen first shows it.
    static func of(_ recipe: Recipe) -> RecipeDraft {
        RecipeDraft(
            name: recipe.name,
            yield: recipe.yield ?? "",
            prepTime: recipe.prepTime ?? "",
            cookTime: recipe.cookTime ?? "",
            totalTime: recipe.totalTime ?? "",
            ingredientsText: recipe.ingredients.joined(separator: "\n"),
            instructionsText: recipe.instructions.joined(separator: "\n"),
            image: recipe.image ?? ""
        )
    }

    /// One entry per non-blank line, trimmed. Split on "\n" only, like Kotlin's `split('\n')`
    /// (a "\r\n" pair is one Character in Swift, so split on the scalar instead).
    static func lines(_ text: String) -> [String] {
        text.unicodeScalars.split(separator: "\n", omittingEmptySubsequences: false)
            .map { trimmed(String(String.UnicodeScalarView($0))) }
            .filter { !$0.isEmpty }
    }

    /// Kotlin's `trim()`: drops leading and trailing whitespace and control characters.
    private static func trimmed(_ s: String) -> String {
        s.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func optional(_ s: String) -> String? {
        let t = trimmed(s)
        return t.isEmpty ? nil : t
    }
}
